### ISSUE TITLE
Support Android versions before 5

### ISSUE THREAD

**Issue by M66B** [M66B/NetGuard#243]
[VpnService](http://developer.android.com/reference/android/net/VpnService.html) is available from SDK 14 = Ice Cream Sandwich.

**Comment by licaon-kter** [comment]
Xperia 4.0.4 (API Level: 15), crash on start: https://gist.github.com/licaon-kter/a09dacf446c8606108a5

**Comment by M66B** [comment]
Reported on XDA:

```
02-01 05:01:08.356 24772 24772 I dalvikvm: Could not find method android.net.ConnectivityManager.getAllNetworks, referenced from method eu.faircode.netguard.Util.getNetworkInfo 02-01 05:01:08.356 24772 24772 I dalvikvm: Could not find method android.telephony.SubscriptionManager.from, referenced from method eu.faircode.netguard.Util.getSubscriptionInfo 02-01 05:01:08.366 24772 24772 I dalvikvm: Could not find method android.content.Context.checkSelfPermission, referenced from method eu.faircode.netguard.Util.hasPhoneStatePermission 02-01 05:01:08.366 24772 24772 I dalvikvm: Could not find method android.os.PowerManager.isInteractive, referenced from method eu.faircode.netguard.Util.isInteractive 02-01 05:01:08.366 24772 24772 I dalvikvm: Could not find method android.telephony.SubscriptionManager.from, referenced from method eu.faircode.netguard.Util.isInternational 02-01 05:01:08.376 24772 24772 D dalvikvm: Trying to load lib /data/app-lib/eu.faircode.netguard-1/libnetguard.so 0x43757430 02-01 05:01:08.386 24772 24772 E dalvikvm: dlopen("/data/app-lib/eu.faircode.netguard-1/libnetguard.so") failed: dlopen failed: cannot locate symbol "sigemptyset" referenced by "libnetguard.so"... 02-01 05:01:08.386 24772 24772 W dalvikvm: Exception Ljava/lang/UnsatisfiedLinkError; thrown while initializing Leu/faircode/netguard/Util; 02-01 05:01:08.406 24772 24772 E AndroidRuntime: Process: eu.faircode.netguard, PID: 24772 02-01 05:01:08.406 24772 24772 E AndroidRuntime: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "sigemptyset" referenced by "libnetguard.so"... 02-01 05:01:08.406 24772 24772 E AndroidRuntime: at eu.faircode.netguard.Util.<clinit>(SourceFile:76) 02-01 05:01:08.406 24772 24772 E AndroidRuntime: at eu.faircode.netguard.ApplicationEx.onCreate(SourceFile:33) 02-01 05:01:08.416 599 3230 W ActivityManager: Force finishing activity eu.faircode.netguard/.ActivityMain
```

**Comment by licaon-kter** [comment]
Apparently building with a newer NDK, some functions that previously were implemented as inline functions (say before android-21), now exist as proper separate standalone functions.
So, maybe add that implementation somewhere like:

```
int sigemptyset(sigset_t *set)
{
memset(set, 0, sizeof *set);
return 0;
}
```

**Comment by M66B** [comment]
Good research!
Where did you find this?

**Comment by licaon-kter** [comment]
On the internet, several sources like:
https://github.com/android/platform_bionic/blob/master/libc/bionic/sigemptyset.cpp
https://code.google.com/p/android/issues/detail?id=73725
https://forum.openframeworks.cc/t/android-master-branch/20106/5

**Comment by licaon-kter** [comment]
Xperia 4.0.4, 0.80-beta4:
* on boot it starts, it's in memory, the toggle is on, but the VPN is not up
* save About logcat (in state like above), calls OI, then also lists other handlers like ES Explorer, but file ends up empty
* export yields OI and saves ok though (BTW, this has an export that works: https://github.com/nolanlawson/Catlog maybe look a bit at it)
* export does not save with a date in it's name so a new export overwrites the old one (without a notice)
* import yields OI and loads ok
* log view, titlebar back arrow does nothing (but the back arrow works on the Pro features view works!)
* settings view does not have a titlebar back arrow 
* dialogues look strange, they need a margin on the left maybe
* notifications are not expandable (was added in later Android versions) BUT I hope the app does not update the address access in the background (it should not, maybe this can conserve some battery)
* filter setting is missing (is it normal!?)
* had a strange situation: app was allowed on Wi-Fi (no SIM card in device), but the log was registering the addresses as blocked on mobile.
* all traffic is blocked or something, see the point above too, I'll test some more

**Comment by M66B** [comment]
"*export does not save with a date in it's name so a new export overwrites the old one (without a notice)*": that is correct, since only the directory can be picked.

"*filter setting is missing (is it normal!?)": filtering needs to be enabled always on Android <5 and not to confuse users this settings is being removed in this case.

What is "*disconnecting the VPN*" ?

**Comment by M66B** [comment]
I can reproduce the VPN not starting on boot. It is caused by a bug in Android :-(
(Android does not remember the VPN was approved by the user)

**Comment by M66B** [comment]
The logcat works fine for me. Could you please check the logcat from your PC for error messages regarding this?

**Comment by M66B** [comment]
The dialogs look okay for me on KitKat 4.4.4.
Can you please provide screenshots?

**Comment by M66B** [comment]
I cannot reproduce the app being blocking on mobile while on Wi-Fi.

**Comment by M66B** [comment]
I can also not reproduce everything blocked on first install.

**Comment by M66B** [comment]
Disabling NetGuard disables the VPN for me.
There could be going something wrong, but I need to see a logcat to check things.

**Comment by M66B** [comment]
"*disconnecting the VPN might make the app crash and not start*": logcat needed as well.

**Comment by M66B** [comment]
"*blocked apps yield a notification on access too*" that is by design, so you can check if not something needs to be allowed to make an application work.

So, block application using the icons. Run application. Allow only hosts you want to allow.

**Comment by licaon-kter** [comment]
> The dialogs look okay for me on KitKat 4.4.4.
https://imgur.com/a/Lt49w

> I cannot reproduce the app being blocking on mobile while on Wi-Fi.
Was fixed after reboot.

> that is by design, so you can check if not something needs
> to be allowed to make an application work.
I know the reason (that's why I asked this initially anyway), but I did not see that in 0.80-beta3.

Logcat ( via CatLog): https://gist.github.com/licaon-kter/06b179d8baecfaa2bfed
Has all the things I mentioned (logcat error, netguard stopped, vpn stopped etc)

Why bother opening/closing this all the time? :)

**Comment by M66B** [comment]
The logcat should be fixed now.

Android is trying to kill the background service because of an ANR.

```
02-02 06:28:40.627 W/ActivityManager(1814): Timeout executing service: ServiceRecord{4191f4c8 eu.faircode.netguard/.SinkholeService}
02-02 06:28:41.968 E/ActivityManager(1814): ANR in eu.faircode.netguard
02-02 06:28:41.968 E/ActivityManager(1814): Reason: Executing service eu.faircode.netguard/.SinkholeService
```

I have no idea why :-(

**Comment by M66B** [comment]
Both *onCreate* and *onStartCommand* do not significant work, so I have no explanation for an ANR.
Unfortunately there is logging missing before the ANR occurs, so maybe you can try to capture a logcat earlier.

**Comment by licaon-kter** [comment]
That's the whole logcat as exported by CatLog, will de adb one have even more info?

**Comment by M66B** [comment]
The adb logcat will have more info, since you cannot start CatLog at boot time.
The logcat buffer has a limited size, so a limited history as well.

**Comment by licaon-kter** [comment]
In the mean time, ANR: https://github.com/SalomonBrys/ANR-WatchDog

Steps:

**Comment by M66B** [comment]
"*VPN is NOT destroyed*" this is something I cannot reproduce, but I love to see the logcat about it.
This could be the root cause of the ANR.

**Comment by licaon-kter** [comment]
Logcat after I did those steps (still from CatLog): https://gist.github.com/licaon-kter/db12e11e305e8fd63bf5

**Comment by M66B** [comment]
Can you please set the native log level to 'debug' and capture a logcat again?
It seems native_stop is hanging.

```
02-02 10:47:02.498 I/NetGuard.Service(21423): Revoke
02-02 10:47:02.518 I/Vpn     (1814): Switched from eu.faircode.netguard to [Legacy VPN]
02-02 10:47:02.528 I/NetGuard.Service(21423): Destroy
02-02 10:47:02.528 I/NetGuard.Service(21423): Stop native clear=true
02-02 10:47:02.528 W/NetGuard.JNI(21423): Stop tun 255 clear 1
```

**Comment by licaon-kter** [comment]
In CatLog it's on Verbose (by default) does that not cover Debug too?
In NetGuard I'll need beta 5 build. :)

**Comment by M66B** [comment]
CatLog cannot capture what NetGuard doesn't log, that is why you need to lower the native log level.

I am testing ICMP ATM, if it is finished, I will build a new version for you.

**Comment by licaon-kter** [comment]
They're both on Verbose, sorry if I missed to mention that.

**Comment by M66B** [comment]
Looking again in the provide logcat:

```
02-02 10:47:02.528 W/NetGuard.JNI(21423): Stop tun 255 clear 1
```

Will always be followed by either *Kill thread xxx* (debug) or *Not running* (warning).
I don't see 'not running', so my conclusion is native log level was above debug, else 'kill thread' would have been appeared.

Please try this again with native log level set to 'debug' in the development settings, so we can see what is happening (or not).

**Comment by M66B** [comment]
Looking into the previous log provide again I found:

```
protect socket failed
```

This is pretty fatal and will result in all traffic being blocked.

However I found a normal shutdown sequence too:

```
02-02 09:29:27.597 W/NetGuard.JNI(14552): Stop tun 180 clear 1
02-02 09:29:27.597 D/NetGuard.JNI(14552): Kill thread 7a1568
02-02 09:29:27.597 D/NetGuard.JNI(14552): Join thread 7a1568
```

Followed by 'Force stopping package eu.faircode.netguard uid=10148' (manual stop?)

**Comment by M66B** [comment]
Test version with potential fix for destroy problem and else with extra logging:
http://www.netguard.me/NetGuard-v0.80-beta-4-2016020101-all-release.apk

**Comment by M66B** [comment]
If you download the same test release again the layout (padding) should be fixed too (another Android Studio glitch).

**Comment by licaon-kter** [comment]
Maybe later... anyway my device is this: http://www.gsmarena.com/sony_xperia_sola-4408.php running official (albeit rooted) Sony ICS 4.0.4

With the (bad layout) build and Debug level in NetGuard and Verbose in CatLog:
* logcat: https://gist.github.com/licaon-kter/61e9cea57393688dfbc4
* ANR traces: https://gist.github.com/licaon-kter/0d9b86c18c39433e571b

Issues still present:
* had to reboot after install, as everything seemed blocked (did not killed it from memory first, I kinda expect the system to do it on upgrade right?)
* logcat saving still 0bytes
* NetGuard stopping still not destroying VPN
* destroying VPN via system yields a NetGuard notification (exclamation mark in circle: "was stopped by" whatever that says since I can't expand it) then a black screen and an ANR in NetGuard.

Works:
* being asked on boot to re-enable NetGuard
* arrows

**Comment by M66B** [comment]
I think this is causing the hang issue:
https://bugzilla.mozilla.org/show_bug.cgi?id=1093893

**Comment by licaon-kter** [comment]
So you need to add special case too?

**Comment by M66B** [comment]
Just done ;-)
Building a test version for you right now.

**Comment by M66B** [comment]
You can download and try the workaround (same link as before).

**Comment by licaon-kter** [comment]
First thing I see, the "app is disabled, use the toggle above this text" dissapears on NetGuard toggle on BUT the app list does NOT use that space, it's just empty there above all apps. Pulled to refresh several times to no avail. Going to Settings an back fixed it. Can't repro it, even by kill/restart now works ok.

Toggle app of still not destroying VPN, BUT destroying VPN does not ANR NetGuard anymore, yay. :)

Notification might need a case, do write allowed/blocked/hosts text after "limits active", because on 4.x you can't expand it. 

Settings `Wi-Fi&mobile block -> ON` and `allow both with screen on->OFF` yields `main Wi-Fi&mobile->off BUT both screen on->ON` for all apps and traffic *is* allowed for all.

**Comment by M66B** [comment]
Please provide another logcat, so I can check things.

licaon-kter <notifications@github.com> schreef op 2 februari 2016 18:44:06 CET:

**Comment by licaon-kter** [comment]
That text on boot is not nice, I can see about half of it in the notification. :(
After the reboot disabling NetGuard destroys the VPN correctly, yay. :)

So ICS or this ROM has a thing for reboots. :-\

Logcat still 0 bytes, CatLog says: https://gist.github.com/licaon-kter/8598ed43f3009d2b7510

**Comment by M66B** [comment]
"*That text on boot is not nice*": which text do you exactly mean?

"*After the reboot disabling NetGuard destroys the VPN correctly*": so my workaround in native code is okay? (looking into the provide logcat it seems it does) Does starting and stopping using the switch work okay?

Can you check the logcat (with CatLog) after trying to write a logcat with NetGuard?

**Comment by licaon-kter** [comment]
The notification (that announces you that NG could not autostart at boot) has half the text, Android limitation since you can't expand it.

Yes, stops/stars as expected, VPN state is controlled correctly.

I did all the tests, so if it's not visible in that logcat I don't have any other means to log it.

**Comment by M66B** [comment]
I am not sure what to do about the notification, since this is mainly an Android limitation.

**Comment by licaon-kter** [comment]
Shorten it: "Please enable NetGuard"

Also, as I mentioned, main notification on sdk=15, replace `msg_started` with `x blk, y allw, z hosts`

**Comment by licaon-kter** [comment]
0.86-beta2 on Xperia 4.0.4

Issues:
* logcat still 0 bytes
* a crash
* mobile device, no SIM-card inside, setting up a home network w/ or w/o `Handle metered Wi-Fi networks` checked always sends traffic to mobile data ending up as blocked.

Logcat via CatLog: https://gist.github.com/licaon-kter/2d8c1bf9bde79799f13b

**Comment by M66B** [comment]
* Logcat: Android might not allow/support reading the logcat, the version below will log the number of lines retrieved from Android, so we can check if this is the case
* Ugly crash, which should be fixed with the version below
* Metered networks: I see no proof of this in the logcat

http://www.netguard.me/NetGuard-v0.86-beta-2016020902-all-release.apk

**Comment by licaon-kter** [comment]
`Logcat lines=0`

I have *OI Explorer* from F-Droid if that matters, also sending the log out of Bluetooth fails too, although I can send files fine from *ES Explorer* otherwise.

Anyway, proof or not, as soon as I choose one home network (currently connected or not: `wifi=true home="MyWifi1" network=MyWifi1 metered=true`) the log shows blocked access on mobile and I can't browse the internet.

Funny thing, had https://github.com/ppareit/swiftp running, and while Chrome would register as blocked in mobile data and could not access the internet ( `... IP syn 1 not allowed` ), the FTP server was accessed successfully on local LAN, also it was not registering in the NetGuard traffic log at all but the logcat has the session.

Logcat via CatLog: https://gist.github.com/licaon-kter/a94928a4de4bd9f79658

**Comment by M66B** [comment]
```
02-10 16:22:50.405 I/NetGuard.Main(15441): Preference wifi_homes=["MyWifi1"]
02-10 16:22:50.515 I/NetGuard.Service(15441): Not at home
02-10 16:22:50.515 I/NetGuard.Service(15441): Get allowed connected=true wifi=true home="MyWifi1" network=MyWifi1 metered=true generation=null roaming=false interactive=true tethering=false filter=true
```

```
        if (wifi && ssidHomes.size() > 0 && !ssidHomes.contains(ssidNetwork)) {
            metered = true;
            Log.i(TAG, "Not at home");
        }
```
Something is indeed going wrong, but I don't see what.

You have set one home network: *wifi_homes=["MyWifi1"]*
You are connected on wifi to MyWifi1: *connected=true wifi=true ... network=MyWifi1*
Which is marked as metered, but the condition above should be false :-(

```
wifi && ssidHomes.size() > 0 && !ssidHomes.contains(ssidNetwork)
```

*wifi=true* (says 'Get allowed')
*ssidHomes.size() > 0* (there is one set in the preferences)
*!ssidHomes.contains(ssidNetwork)* ('MyWifi1' in 'Get allowed' is in the list

**Comment by M66B** [comment]
Maybe I have an explanation: is 'MyWifi1' a metered network?
In this case disable the setting '*Handle metered Wi-Fi networks*'.

**Comment by M66B** [comment]
The remaining, solvable issue is not about Android versions before 5 anymore, so I am going to close this issue.

If the above explanation isn't the correct explanation, please create a new issue.
